### PR TITLE
Adding ionization capabilities for viscous simulations (NEMO)

### PR DESCRIPTION
**Pull request by WallyMaier**
## Proposed Changes
This PR adds the capability to simulate the AIR-7 gas model using SU2TCLib for viscous problems. Specific contributions include:
- Fix Gupta-Yos transport models for all flows (including ionized)
- inclusion of Coulomb interaction cross sections for ion-ion collisions

Following the implementation detailed in Scalabrin's thesis.

## Related Work
PRs #1497 #1555 

Update to NEMO documentation includes description of GY model, see: https://github.com/su2code/su2code.github.io/pull/116

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:1027]: @jtneedels these values are undefined in Scalabrin's thesis.  Something to look into

Comment (jtneedels) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:1027]: I believe these are detailed in Table A.10 where attractive/repulsive Coulomb potential is discussed. Since these are all ion-ion pairs, I think the collision cross section is based on Coulomb potential since all are either attractive/repulsive.

**Comment by jtneedels** [comment]
Reading Scalabrin, he says the WBE transport model breaks down at Ttr > 10,000 K, and he uses the GY model for incorporating ionization in LeMANS. So I think the roadmap to for viscous/ionized flows is:

1. Implement Coulomb potential for ion-ion collisions
2. Implement Gupta-Yos model according to Scalabrin
3. Add some flag to make sure transport model for viscous-ionized is always GY

**Comment by WallyMaier** [comment]
The path forward sounds good.  Gupta-Yos needs to be reworked/fixed :(

**Comment by jtneedels** [comment]
> The path forward sounds good. Gupta-Yos needs to be reworked/fixed :(

Yeah, I'm working on that now. I think one way to do this would be to code in whether to use the Coulomb potential using the omega00 tables for each collision pair. For example, if omega00(ispecies, jspecies, 0) == 1.0, then use Coulomb. We could use 1.0, or some other non-physical value.

**Comment by jtneedels** [comment]
An update: I re-worked this so the computation of the delta_ij values (used in each of the GY functions) is itself a function to remove repeated code. It is now also set up to handle ionized flows. Use of the Coulombic potential will signaled by the omega00 and omega11 values. Forward work is ensuring the correct collision integral curve fit values are used, modifying the omega00 and omega11 tables.

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:1949]: So this would mean that Gupta-Yos must be used with ionization.
Would there be any benefit of allowing neutral simulations with Gupta-Yos?

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:1954]: Good use of these functions.  Maybe we could be more specific with the name "ComputeCollisionalDeltas"?   Just brain storming.

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:2002]: 👍

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:1949]: I think Gupta-Yos should be computationally efficient, though I don't know which model is more accurate at "low" temperatures

Comment (jtneedels) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:1949]: We can make this more general by saying if (ionization && iSpecies == 0), so that if there is not ionization it uses Ttr vs Tve. That would give the option of using GY for neural sims.

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (jtneedels) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:1954]: Sure, yes that sounds good!

**Comment by WallyMaier** [comment]
I dont believe there is any other places to implement changes for viscous cases.

File: Common/src/CConfig.cpp
Comment (WallyMaier) [inline Common/src/CConfig.cpp:6158]:
```suggestion
            SU2_MPI::Error("The GAS_MODEL given is unavailable using CSU2TCLIB. Choose one of the options: N2, AIR-5, AIR-7, or ARGON.", CURRENT_FUNCTION);
```

File: Common/src/CConfig.cpp
Comment (WallyMaier) [inline Common/src/CConfig.cpp:6165]:
```suggestion
          SU2_MPI::Error("The GAS_MODEL given is unavailable using CSU2TCLIB. Choose one of the options: N2, AIR-5, AIR-7, or ARGON.", CURRENT_FUNCTION);
```

File: SU2_CFD/include/fluid/CSU2TCLib.hpp
Comment (WallyMaier) [inline SU2_CFD/include/fluid/CSU2TCLib.hpp:276]:
```suggestion
   *\param[out] Delta_1 or Delta_2 - Collision term between species i and j
```

**Comment by jtneedels** [comment]
Update: model runs with air-5 and air-7 but diverges in ~10 iterations. Need to check implementation, look for bugs.

**Comment by jtneedels** [comment]
For any folks following this saga, one of the issues was thermal conductivity. I think I figured it out and it’s a weird one… In short: it was the thermal conductivities, but not for the reason we think. They are divided by a denom term that depends on a value a that indeed goes to nan if any free-stream concentrations are zero. However, the reason it doesn’t diverge immediately is if that denom term is less than eps, the thermal conductivity is set to zero, so thermal conductivity was always zero which is why it blew up after a few iters rather than the first one.
Fix: FS concentrations > 0, but also have to remove the if-statement that sets k_tr to zero. The denom term is order 1E-21 and always less than eps, so with that in the code it is always 0.

Diffusion coefficient is now the only parameter giving wrong values, so need to check that next.

**Comment by jtneedels** [comment]
Update: Wally found an index issue with the Ds calculation. Now model works for neutral and ionized flows, as long as free-stream concentrations are non-zero. Forward work is to fix it so that free-stream concentrations can be zero and not cause a problem.

**Comment by jtneedels** [comment]
Hi @pcarruscag I think we're at a place in this PR where we can ask for a code review. In particular, I'm wondering if there is a way to reduce the code complexity in the ComputeCollisionDelta() function. Moving this to its own function reduced repeated code (which was called for each GY transport model) but now I've got some cumbersome logic statements and I'm not sure how to simplify.

**Comment by jtneedels** [comment]
Both Gupta-Yos and WBE are now capable of running with ionization for the viscous simulations using explicit time integration. Viscous numerics needs to be modified to permit implicit. Initial testing at Mach 10 and Mach 15 by @WallyMaier show promising results for GY.

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:1894]: Can we add some comments to this block just for clarity?

File: .github/workflows/regression.yml
Comment (WallyMaier) [inline .github/workflows/regression.yml:131]: Just a note to not forget this.

**Review by pcarruscag — APPROVED** [review]
LGTM code-wise

**Comment by jtneedels** [comment]
Status update: Wally ran a Mach 20 cylinder test case with AIR-7 model using SU2_NONEQ and results looked good. I need to check everything works with MPP air-7, and regressions need to be updated. Once confirmed/updated I think we can merge.

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (jtneedels) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:1898]: Sure, but I think this is best explained by the PR to documentation which gives the actual equations.

Comment (WallyMaier) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:1898]: I think something like  'Compute collision term'  and something to help distinguish the loops a little bit.  Nothing crazy.

File: .github/workflows/regression.yml
Comment (jtneedels) [reply .github/workflows/regression.yml:131]: @WallyMaier Anything we still need to do here?

File: SU2_CFD/include/fluid/CSU2TCLib.hpp
Comment (jtneedels) [inline SU2_CFD/include/fluid/CSU2TCLib.hpp:74]: Should I add a comment with a reference for these values, or okay just to keep track of that in documentation?

**Review by jtneedels — APPROVED** [review]

File: .github/workflows/regression.yml
Comment (WallyMaier) [reply .github/workflows/regression.yml:131]: Not until we finally push

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:2063]: @pcarruscag @jtneedels This helps prevent the solution going NaN and seems to have little impact on solution quality/convergence behavior.   Is there a better/safer way to do this?

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:2052]: Is there a better way to check for Inf?   In some cases d1 or d2 can be infinity but this causes NaNs.

File: SU2_CFD/include/fluid/CSU2TCLib.hpp
Comment (WallyMaier) [reply SU2_CFD/include/fluid/CSU2TCLib.hpp:74]: I think that would be helpful!

**Review by jtneedels — APPROVED** [review]

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (jtneedels) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:2063]: LGTM - we can't say if denom < EPS because these values are so small (that was causing issues with the original implementation) so I think this is good.

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:1970]:
```suggestion
        su2double d1_ij = fmin(ComputeCollisionDelta(iSpecies, jSpecies, Mi, Mj, T_col, true), 1E16);
```

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (pcarruscag) [inline SU2_CFD/src/fluid/CSU2TCLib.cpp:2002]: Repeated, why don't you put the limit inside the function?
Or define a constant for the limit, otherwise this repetition is bound to cause problems at some point

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (pcarruscag) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp:2063]: Is denom always positive? If so consider using fmax(denom, EPS) comparing floats with == can cause surprises.

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp]: I moved it into ComputeCollisionDelta function

File: SU2_CFD/src/fluid/CSU2TCLib.cpp
Comment (WallyMaier) [reply SU2_CFD/src/fluid/CSU2TCLib.cpp]: Unfortunately, adding in the fmax causes issues down the line in the simulation.  I added a check for denom<= 0 to avoid any extra weirdness.

**Pull request merged**
